dbms/src/Common/TraceCollector.cpp

```diff
@@ -51,11 +59,98 @@ TraceCollector::~TraceCollector()
         LOG_ERROR(log, "TraceCollector thread is malformed and cannot be joined");
     else
     {
-        TraceCollector::notifyToStop();
+        stop();
         thread.join();
     }
 
-    trace_pipe.close();
+    pipe.close();
+}
+
+void TraceCollector::collect(TraceType trace_type, const StackTrace & stack_trace, int overrun_count)
+{
+    /// Quickly drop if signal handler is called too frequently.
+    /// Otherwise we may end up infinitelly processing signals instead of doing any useful work.
+    ++write_trace_iteration;
+    if (overrun_count)
+    {
+        /// But pass with some frequency to avoid drop of all traces.
+        if (write_trace_iteration % overrun_count == 0)
+        {
+            ProfileEvents::increment(ProfileEvents::QueryProfilerSignalOverruns, overrun_count);
+        }
+        else
+        {
+            ProfileEvents::increment(ProfileEvents::QueryProfilerSignalOverruns, overrun_count + 1);
+            return;
+        }
+    }
+
+    constexpr size_t buf_size = sizeof(char) + // TraceCollector stop flag
+        8 * sizeof(char) +                     // maximum VarUInt length for string size
+        QUERY_ID_MAX_LEN * sizeof(char) +      // maximum query_id length
+        sizeof(UInt8) +                        // number of stack frames
+        sizeof(StackTrace::Frames) +           // collected stack trace, maximum capacity
+        sizeof(TraceType) +                    // trace type
+        sizeof(UInt32) +                       // thread_number
+        sizeof(Int64);                         // size
+    char buffer[buf_size];
+    WriteBufferFromFileDescriptorDiscardOnFailure out(pipe.fds_rw[1], buf_size, buffer);
```

Comment:
Is it time to write a ser/de method for TraceLogElement? Or to even switch a trace from Array to std::array, so that TraceLogElement is fixed size and we can ship it verbatim over the pipe. Actually we can ship it even if it has pointers inside, because we're in the same process -- this would have move semantics. But I don't trust these pipes enough, maybe it'll leak :)

Reply:
> switch a trace from Array to std::array, so that TraceLogElement is fixed size

No need for that. When traces comes frequently, pipe is overflowed.